Epic 8147: WNPRC Purchasing System - Request Data Entry

labkey-bpatel opened the pull request:
#### Rationale
WNPRC Purchasing system's Request Data Entry form for the lab end users.

#### Related Pull Requests
https://github.com/LabKey/ehrModules/pull/111

#### Changes
- Add linked schema template for ehr_billing.aliases
- Implement React panels for Request Order and Line Items
- Add dropdowns for Account to charge, vendor, shipping destination, line item unit
- Implement Other vendor workflow
- Error validation
- Add subtotal and total
- Handle Cancel and Submit for Review
- Create action for saving the request form & related server side changes

labkey-bpatel committed:
Add LinkedSchema template for ehr_billing.aliases. Update metadata.

labkey-bpatel committed:
Render 'Request Order panel' (wip, shows basic form layout for now)

labkey-bpatel committed:
Add dropdown for 'Account to charge'

labkey-bpatel committed:
Move ehr_billingLinked.template.xml from purchasing module to wnprc_billing module

labkey-bpatel committed:
Refactor to use functional component.

labkey-bpatel committed:
Enable remaining Request Order panel inputs. Update sample data.

labkey-bpatel committed:
Handle 'Other' account. Cleanup.
Attempt for 'Other' vendor (commented out for now).

labkey-bpatel committed:
Add Line Items panel. Refactor common reusable code out to appropriate util & action files.

labkey-bpatel committed:
Handle adding of new line item row. Fix side effects on Request Order panel.

labkey-bpatel committed:
Separate out functional units of code in different files for better readability.

labkey-bpatel committed:
Fix issues for the Line Item component. Calculate subtotal.

labkey-bpatel committed:
Update useCallback dependencies.

labkey-bpatel committed:
Reformat code for readibility. Add buttons for adding an item row and deleting an item row. Other minor fixes.

labkey-bpatel committed:
Fix styling

labkey-bpatel committed:
Handle row delete

labkey-bpatel committed:
Styling updates - add scss and remove style attribute. Other minor changes

labkey-bpatel committed:
Handle subtotal correctly. Add Total. More styling.

labkey-bpatel committed:
Use evt.target.checked instead of evt.target.value for the 'Controlled substance' checkbox

labkey-bpatel committed:
Adjust margin

labkey-bpatel committed:
Format dropdowns with additional data.

labkey-bpatel committed:
Format dropdowns with additional data - minor refactor.

labkey-bpatel committed:
Match panel color to LabKey's Look & Feel theme.

labkey-bpatel committed:
Add numeric pattern for unit price and quantity inputs.

labkey-bpatel committed:
New vendor input form upon selection of 'Other' vendor. Format Subtotal and Total with 2 decimal places.

labkey-bpatel committed:
Remove unused text.

labkey-bpatel committed:
Add client side validation on missing required fields upon Submit for Request Order & Line Items panels.

labkey-bpatel committed:
Add client side validation on missing required fields for the new vendor.

labkey-bpatel committed:
Handle New vendor display (for the user to see the entered values, since the pop will disappear after save, and the user has no way of seeing the new vendor entry).

labkey-bpatel committed:
Display New Vendor values, in case the user would like to edit them. Add & handle Edit button.

labkey-bpatel committed:
cleanup

labkey-bpatel committed:
New vendor workflow - redo

labkey-bpatel committed:
New vendor workflow - handle edit vendor correctly.

labkey-bpatel committed:
Refactor to save the rowId of the drop down select. Submit form related updates (wip).

labkey-bpatel committed:
Submit request - purchasing request insert

labkey-bpatel committed:
Submit form related updates

labkey-bpatel committed:
Add other acct and investigator to the overview query

labkey-bpatel committed:
New vendor workflow - fix issues

labkey-bpatel committed:
New vendor workflow - error validation updates

labkey-bpatel committed:
New vendor workflow - update model only if no errors

labkey-bpatel committed:
Save qc state upon new request submit

labkey-bpatel committed:
cleanup

labkey-bpatel committed:
Update 'type' from subtotal input

labkey-bpatel committed:
Handle cancel

labkey-bpatel committed:
Move linked template to purchasing folder

labkey-bpatel committed:
Move linked template to wnprc_billing module. Bug fixing upon save when errors were previously present.

labkey-bpatel committed:
Add vendor to the overview query

labkey-bpatel committed:
Store newly created vendor's id in purchasingRequest table

labkey-bpatel committed:
Remove todo (already implemented)

labkey-bpatel committed:
Update other vendor label

labkey-bpatel committed:
Move eventlistener for dirty state in its own useeffect()

labkey-bpatel committed:
Reset Vendor input when user hits Cancel and doesn't enter a new vendor

labkey-bpatel committed:
Reset dirty state upon submit

labkey-bpatel commented:
@labkey-martyp - can you please review React code. Thanks!

labkey-bpatel commented:
@labkey-jeckels - can you please review Java code. Thanks!

labkey-jeckels commented on WNPRC_Purchasing/resources/queries/ehr_purchasing/purchasingRequestsOverview.sql:
```diff
@@ -1,11 +1,17 @@
-       sum(items.unitCost) AS totalCost
+       round(sum(items.quantity * items.unitCost), 2) AS totalCost
 FROM ehr_purchasing.purchasingRequests pr
 LEFT JOIN ehr_purchasing.lineItems items ON pr.requestId = items.requestId
```
This will be more efficient if you do a subselect on lineItems that groups by just requestId, and then join that with purchasingRequests. Let me know if you want to talk about why this is, but the basic idea is that it reduces the complexity of the GROUP BY

labkey-bpatel replied:
Yup that makes sense, my query was pretty slow as well, this helped! Thank you!

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingManager.java:
```diff
@@ -90,4 +107,112 @@ private void addData(Container c, User user, String schemaName, String tableName
+        //Line items data
+        List<Map<String, Object>> lineItemsData = new ArrayList<>();
+        ObjectMapper mapper = new ObjectMapper();
+        WNPRC_PurchasingController.LineItem[] lineItems = mapper.convertValue(requestForm.getLineItems(), WNPRC_PurchasingController.LineItem[].class);
```
This seems like a lot of gyrations from receiving JSON -> Java objects -> Map<String, Object>. Can we just feed the JSON objects (which should be easy to get as maps) directly into the QueryUpdateService calls?

labkey-bpatel replied:
Agreed! now changed!

labkey-bpatel committed:
Automated test - update setup to include extensible column & linked schema. Adjust sample data. Update existing test.

labkey-martyp commented on WNPRC_Purchasing/package.json:
```diff
@@ -16,9 +16,40 @@
+    "style-loader": "1.3.0",
+    "ts-loader": "8.0.4",
+    "typescript": "4.0.3",
+    "url-loader": "4.1.0",
```
You don't need to import all these that are already imported by @labkey/build.  See [here ](https://github.com/LabKey/labkey-ui-components/blob/9449914a3f49682ac9f677fdfb6729f7f3453d49/packages/build/package.json#L14)for all the packages that @labkey/build imports

labkey-bpatel replied:
Ah, thats good to know! Now changed! Thank you!

labkey-martyp commented on WNPRC_Purchasing/postcss.config.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = (ctx) => ({
+    parser: ctx.parser ? 'sugarss' : false,
+    map: ctx.env === 'development' ? ctx.map : false
+});
```
I believe @cnathe is just merging in a change so you won't need this if you are using @labkey/build

labkey-bpatel replied:
Thats great! Now updated

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+    //equivalent to componentDidMount and componentDidUpdate (if with dependencies, then equivalent to componentDidUpdate)
+    useEffect(() => {
+        // is fired on component mount
+        let isNewRequest = ActionURL.getParameter('isNewRequest');
```
const

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+        let isNewRequest = ActionURL.getParameter('isNewRequest');
+        if (isNewRequest) {
+            getData('core', 'qcState', 'RowId, Label').then(vals => {
+                let idx = vals.findIndex(qcstate => qcstate['Label'] === 'Review Pending');
```
const - Use const whenever you are not planning to change a value

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+        return () => {
+            window.removeEventListener('beforeunload', handleWindowBeforeUnload);
+        }
+    }, [isDirty, setIsDirty]);
```
You don't need to include setState functions like setIsDirty in dependency array.  From the react hook [docs](https://reactjs.org/docs/hooks-reference.html),
 
"React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list."

labkey-bpatel replied:
That's good to know, now done!

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+        window.addEventListener('beforeunload', handleWindowBeforeUnload);
+
+        // is fired on component unmount
+        return () => {
+            window.removeEventListener('beforeunload', handleWindowBeforeUnload);
+        }
```
Nice cleanup

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+        if (isDirty) {
+            event.returnValue = 'Changes you made may not be saved.';
+        }
+    }, [isDirty, setIsDirty]);
```
Same as above about setState functions.  I won't keep writing it but these can be removed from dependency array of any hooks

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+    const lineItemsChange = useCallback((lineItemArray : Array<LineItemModel>)=> {
+        let numOfErrors = 0;
+        for (let i = 0; i < lineItemArray.length; i++) {
+            if (lineItemArray[i].errors && lineItemArray[i].errors.length > 0) {
```
TypeScript 3.7 introduced [Optional Chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining), so you can shorten this with 

`if (lineItemArray[i].errors?.length > 0) {`

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+            }
+        }
+        if (numOfErrors == 0) {
+            setLineItemErrorMsg(undefined);
```
If you use `setLineItemErrorMsg()` it should pass undefined as any parameters

labkey-bpatel replied:
this didn't work, Intellij was throwing an error, so kept as is

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+
+    const lineItemsChange = useCallback((lineItemArray : Array<LineItemModel>)=> {
+        let numOfErrors = 0;
+        for (let i = 0; i < lineItemArray.length; i++) {
```
Using a for loop to count this is fine.  A more elegant way would be to use a reduce function on the array.  Something like (not tested),

`const numOfErrors = lineItemArray.reduce(((count, item) => (lineItemArray[i].errors?.length > 0) ? (count + 1) : count), 0) `

labkey-bpatel replied:
done!

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+    const onCancelBtnHandler = useCallback(() => {
+        setIsDirty(false);
+        const returnUrl = ActionURL.getParameter('returnUrl');
+        window.location.href = returnUrl || ActionURL.buildURL('project', 'begin', getServerContext().container.path);
```
I only see returnUrl used on Cancel.  Should it be used to navigate after submit too?

labkey-bpatel replied:
For now, it is only set on Cancel, for Submit we want the user to go to the Overview grid. But this might change later, so keeping it as is.

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+    const onSaveBtnHandler = useCallback(() => {
+
+        //if required values are not provided, then set errorMessage for the panel and errors on each field to highlight boxes in red
+        let msg = "Unable to submit request, missing required field(s).";
```
Can you make this last word plural based on the count of requestOrderErrors?

labkey-bpatel replied:
Good idea, done!

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+
+    return (
+        <>
+            <RequestOrderPanel onInputChange={requestOrderModelChange} model={requestOrderModel} errorMsg={requestOrderModel.errorMsg}/>
```
Probably don't need to pass errorMsg as a separate prop right?  If it's part of the model prop

labkey-bpatel replied:
doh! good catch! now done!

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+            <RequestOrderPanel onInputChange={requestOrderModelChange} model={requestOrderModel} errorMsg={requestOrderModel.errorMsg}/>
+            <LineItemsPanel onChange={lineItemsChange} lineItems={lineItems} errorMsg={lineItemErrorMsg}/>
+            <button className='btn btn-default' id='cancel' name='cancel' onClick={onCancelBtnHandler}>Cancel</button>
+            <button className='btn btn-primary' style={{float:'right'}} id='submitForReview' name='submitForReview' onClick={onSaveBtnHandler}>Submit for Review</button>
```
Bootstrap has a class .pull-right you should be able to use here instead of inline style

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -0,0 +1,60 @@
+            columns: colNames,
+            // filterArray: [
+            //     Filter.create()
+            // ]
```
Delete?

labkey-bpatel replied:
deleted

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -0,0 +1,60 @@
+            // filterArray: [
+            //     Filter.create()
+            // ]
+            success: function (results) {
```
You might be able to use Utils.getCallbackWrapper from @labkey/api to get rows for results and useful for a failure callback which would be good here.

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -0,0 +1,60 @@
+                comments: requestOrder.comments,
+                qcState: requestOrder.qcState,
+                lineItems: lineItems,
+                hasNewVendor: (requestOrder.vendor === 'Other' && VendorModel.getDisplayVersion(requestOrder.newVendor)) ? true : false,
```
Not sure if VendorModel.getDisplayVersion(requestOrder.newVendor) can ever be 0.  That will evaluate as false here

labkey-bpatel replied:
this is changed to use !!

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemRow.tsx:
```diff
@@ -0,0 +1,90 @@
+    const onValueChange = useCallback((colName, value) => {
+        const updatedModel = produce(model, (draft: Draft<LineItemModel>) => {
+            draft[colName] = value;
+            if (model.errors && model.errors.length > 0) {
```
`if (model.errors?.length > 0)`

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemRow.tsx:
```diff
@@ -0,0 +1,90 @@
+                <Col xs={4}>
+                    <DescriptionInput
+                    value={model.item}
+                    hasError={model.errors && model.errors.find((field) => field.fieldName === 'item')}
```
`model.errors?.find((field) => field.fieldName === 'item')`

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemRow.tsx:
```diff
@@ -0,0 +1,90 @@
+                      id={'delete-line-item-row-' + rowIndex} title={'Delete item'} className="delete-item-icon"
+                      onClick={onDeleteRow}
+                    >
+                        <FontAwesomeIcon className='fa-faTimesCircle' icon={faTimesCircle} color={'gray'}/>
```
Looks like color is being set in scss too. You should just set color in one place, scss file would allow both hover and regular states

labkey-martyp commented on WNPRC_Purchasing/src/client/components/PurchasingFormInput.tsx:
```diff
@@ -0,0 +1,54 @@
+    colSize?: number;
+}
+
+export class PurchasingFormInput extends React.PureComponent<PurchasingFormInputProps, any>
```
Might be good to be consistent with using FC with hooks or component classes.  At least within the same project

labkey-bpatel replied:
good point, now done

labkey-martyp commented on WNPRC_Purchasing/src/client/components/RequestOrderPanel.tsx:
```diff
@@ -0,0 +1,113 @@
+            <Form>
+                <AccountInput
+                    value={model.account}
+                    hasError={model.errors && model.errors.find((field) => field.fieldName === 'account')}
```
`model.errors?.find`

labkey-martyp commented on WNPRC_Purchasing/src/client/components/VendorInputModal.tsx:
```diff
@@ -0,0 +1,279 @@
+        const updatedModel = produce(updatedNewVendor, (draft: Draft<VendorModel>) => {
+            if (errors.length > 0) {
+                draft.errors = errors;
+                let msg = 'Please fix error(s) before saving:  ';
```
Plural if errors.length > 1?

labkey-martyp commented on WNPRC_Purchasing/src/client/components/VendorInputModal.tsx:
```diff
@@ -0,0 +1,279 @@
+
+    const onInputChange = useCallback((colName, value) => {
+        const updatedVendorModel = produce(updatedNewVendor, (draft: Draft<VendorModel>) => {
+            if (updatedNewVendor.errors && updatedNewVendor.errors.length > 0) {
```
updatedNewVendor.errors?.length > 0

labkey-martyp commented on WNPRC_Purchasing/src/client/components/VendorInputModal.tsx:
```diff
@@ -0,0 +1,279 @@
+    }, [required, column, columnTitle, value, onChange]);
+
+    return (
+        <div style={{width:'100%'}}>
```
We should try to use css classes instead of inline styling

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/model.ts:
```diff
@@ -0,0 +1,110 @@
+        Object.assign(this, values);
+    }
+
+    static create(raw: any, defaultSettings = null): RequestOrderModel {
```
Can you make the raw parameter optional on all these models or at least have a default value?  There are a lot of models created like RequestOrderModel.create({}) instead of RequestOrderModel.create().  The empty object parameter isn't really necessary.

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -13,24 +13,163 @@
+
+export const App : FC = () => {
+
+    const [requestOrderModel, setRequestOrderModel] = useState<RequestOrderModel>(RequestOrderModel.create({}));
```
See my comment in the models.ts about RequestOrderModel.create({})

labkey-martyp reviewed (changes requested):
Overall looks great.  A lot of little nit picky things in comments

labkey-bpatel committed:
Fix reset vendor input on cancel

labkey-bpatel committed:
Code review changes
- Update overview query to use subselect on lineItems
- Use JSONObject instead of Object
- Remove redundant dependencies in package.json
- Remove postcss.config.js, now availabe as part of @labkey/build
- Use const
- Remove setState methods from useEffects dependency list
- Use optional chaining
- Make errors message singular/plural based on num. of errors
- Make raw parameter optional for model's create()
- Cleanup & other minor formatting changes

labkey-bpatel committed:
Remove setstate fn from hooks' dependency array

labkey-sweta committed:
WNPRC_Purchasing : Automation test

labkey-sweta committed:
Merge remote-tracking branch 'origin/fb_wnprc_purchasing_epic8148' into fb_wnprc_purchasing_epic8148

labkey-bpatel commented on WNPRC_Purchasing/test/src/org/labkey/test/tests/wnprc_purchasing/WNPRC_PurchasingTest.java:
```diff
@@ -192,12 +192,80 @@ public void preTest()
+        requestPage.setAccountsToCharge("acct100")
+                .setVendor("Real Santa Claus")
+                .setBusinessPurpose("Holiday Party")
+                .setSpecialInstructions("Ho Ho Ho")
```
Haha! love it!

labkey-bpatel commented on WNPRC_Purchasing/test/src/org/labkey/test/tests/wnprc_purchasing/WNPRC_PurchasingTest.java:
```diff
@@ -192,12 +192,80 @@ public void preTest()
+
+        log("Verifying the validation for mandatory fields");
+        requestPage.submitForReview();
+        checker().verifyEquals("Invalid error message", "Unable to submit request, missing required field(s).", requestPage.getAlertMessage());
```
This is now pluralized based on no. of errors:
So, if one error then "Unable to submit request, missing required field."
More than one then "Unable to submit request, missing required fields."

labkey-sweta replied:
will correct the error message.

labkey-bpatel replied:
ive updated it, since im on this fb

labkey-bpatel committed:
Test update - change error msg

labkey-bpatel committed:
Disable buttons on submit. Cleanup.

labkey-martyp reviewed (approved):
UI side looks good

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_wnprc_purchasing_epic8148

labkey-jeckels reviewed (approved):
Thanks for the update, looks good.

labkey-bpatel commented:
WNPRC Purchasing tests all pass, merging..
